presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -963,7 +963,9 @@ else if (criteria instanceof JoinOn) {
                 // we do it further down when after we determine which subexpressions apply to left vs right tuple)
                 ExpressionAnalyzer analyzer = ExpressionAnalyzer.create(analysis, session, metadata, sqlParser, accessControl);
 
-                Type clauseType = analyzer.analyze(expression, output);
+                // need to register coercions in case when join criteria requires coercion (e.g. join on char(1) = char(2))
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output);
```

Comment:
`canonicalizeExpression` is already a plan optimizer. I believe it had to be called from visitJoin in order to be able to optimize the expression prior to extracting equality conditions, etc. 

BTW, `visitJoin` currently doing too much. Now that we can represent join conditions with arbitrary expressions, `visitJoin` should just create the join node and leave it up to optimizers to extract/simplify/move conditions around later. I started looking into this a couple of months ago, but didn't make much progress.